src/Fizzics/level2.js

```diff
@@ -73,90 +145,81 @@ var FizzicsLevel2 = GObject.registerClass({
         GObject.signal_handler_block(this._model, this._notifyHandler);
 
         try {
-            Object.getOwnPropertyNames(scope).forEach(prop => {
-                const property = this._toModelName(prop);
-                if (!(prop in scope) || scope[prop] === this._model[property])
-                    return;
-                this._model[property] = scope[prop];
-            });
+            this._updateModelFromScope(scope);
         } finally {
             GObject.signal_handler_unblock(this._model, this._notifyHandler);
         }
     }
 
-    _errorRecordAtAssignmentLocation(variable, message, default_value) {
+    _errorRecordAtAssignmentLocation(variable, message, value) {
         const {start, end} = this._codeview.findAssignmentLocation(variable);
-        return {start, end, message, fixme: String(default_value)};
+        return {start, end, message, fixme: String(value)};
     }
 
-    _searchCodeForErrors(scope) {
-        const errors = [];
-
-        this._getModelProps().forEach(pspec => {
-            const propName = this._toScopeName(pspec.get_name());
-            const propType = typeof pspec.default_value;
-            if (!scope[propName])
+    _searchForErrorWithProps(errors, scope, props, prefix) {
+        Object.keys(props).forEach(key => {
+            if (scope[key] === null)
                 return;
-            if (scope[propName] !== null && typeof scope[propName] !== propType) {
+            const modelProp = props[key];
+            const modelType = typeof this._model[modelProp];
+            if (typeof scope[key] !== modelType) {
                 errors.push(this._errorRecordAtAssignmentLocation(
-                    propName,
-                    `Unknown value ${scope[propName]}: value must be a ${propType}`,
-                    pspec.default_value
+                    prefix ? `${prefix}${key}` : key,
+                    `Unknown value ${scope[key]} for ${key}: value must be a ${modelType}`,
+                    this._model[modelProp]
                 ));
             }
         });
+    }
 
+    _searchForErrors(scope) {
+        const errors = [];
+        const props = this._getPropsForGlobals();
+        this._searchForErrorWithProps(errors, scope, props, null);
+        Array.from({length: OBJECTS_NUM}).forEach((value, index) => {
+            const objProps = this._getPropsForIndex(index);
+            this._searchForErrorWithProps(
+                errors, scope.objects[index], objProps, `objects[${index}].`);
+        });
         return errors;
     }
 
+    _generateCodeForIndex(index) {
+        let code = `
+// ${index}
+`;
+        const props = this._getPropsForIndex(index);
+        Object.keys(props).forEach(key => {
+            let model_value = this._model[props[key]];
+
+            /* handle UI customizations here */
+            if (key === 'lock')
+                model_value = !model_value;
```

Comment:
Same comment than above about not understanding what this achieves; I also wonder if we can refactor things so that this enumeration of properties and value changes is only done in a single place.